README.md
[![ci-cd](https://github.com/UBC-MDS/nameforme/actions/workflows/ci-cd.yml/badge.svg)](https://github.com/UBC-MDS/nameforme/actions/workflows/ci-cd.yml) 
[![docs](https://readthedocs.org/projects/nameforme/badge/?version=main)](https://readthedocs.org/projects/nameforme/badge/?version=main)
[![PyPI version](https://badge.fury.io/py/nameforme.svg)](https://badge.fury.io/py/nameforme)
[![codecov](https://codecov.io/gh/UBC-MDS/nameforme/branch/main/graph/badge.svg?token=vKUSBGopdL)](https://codecov.io/gh/UBC-MDS/nameforme)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

# nameforme

A helper python package that can be used to generate names based on the [`dateset`](https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2022/2022-03-22/babynames.csv). This could be used to come up with baby names, character names, pseudonyms, etc. 

Source Data: Contains baby names born in the United States for each year from 1880 to 2017, and the number of children of each sex given each name. Names must appear at least 5 times in each year to be included. (Source: http://www.ssa.gov/oact/babynames/limits.html)

This package is similar to the existing [names](https://pypi.org/project/names/) package by Trey Hunner (last updated in 2014), however, our uses a more recent dataset (with names up to 2017), and more options for users to customize what type of names to generate, including the ability to generate similar sounding names.

You can check the online documentation of this package on [readthedocs](https://nameforme.readthedocs.io/en/main/index.html).

## Features
Note that the name of the functions is not finalized. They are subject to change.

The package is an assimilation of four independent functions:

- `find_unisex_name`: Generate a random set of 10 suggested neutral baby names based on the given limitation and baby names in the past years.

- `find_old_name`: Generate a random set of 10 suggested neutral(by default) baby names based on the given time period and sex.

- `find_similar_name`: Generate a random set of 10 suggested similar baby names based on the syllable of the input name. 

- `find_name`: Generate a random set of 10 suggested baby names based on the given limitations.

## Installation

```bash
$ pip install nameforme
```

## Usage

Below is a basic example of how to use each of the four functions included in this package.

```
# Load all required package functions
from nameforme.nameforme import find_name
from nameforme.nameforme import find_old_name
from nameforme.nameforme import find_similar_name
from nameforme.nameforme import find_unisex_name

# Generate a random set of 10 suggested baby names based on the given limitations.
# if the given limitation can match to at least 10 names, a list of 10 names will be provided
find_name("F", "A", length=3)
#if the given limitation can only match less than 10 names, all matched names will be provided
find_name("m", "b", length=9)

# Generate a random set of suggested neutral(by default) baby names based on the given time period and sex.
find_old_name('1980s', limit=3)

# Generate a random list of names that sound similar to a given user input name.
find_similar_name('Daniel', limit=20)

# Generate the a random set of suggested neutral baby names based on the given limitation and baby names in the past years.
find_unisex_name(bar=0.02,limit=10)
```

## Dependencies
- python = "3.9"
- numpy = "1.24.1"
- pandas = "1.5.3"
- jellyfish = "0.9.0"
- pytest = "7.2.1"

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`nameforme` was created by Daniel Cairns, Eyre Hong, Bruce Wu, Zilong Yi (UBC MDS). It is licensed under the terms of the MIT license.

## Credits

`nameforme` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).

CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.1.0 (2023-01-29)
### Feature
* **new:** Updated readme. ([`c168b59`](https://github.com/UBC-MDS/nameforme/commit/c168b59a67bd1f9c8e5ee078099e39087245e09f))

## v1.0.0 (2023-01-29)
### Feature
* New changes ([`bb076da`](https://github.com/UBC-MDS/nameforme/commit/bb076dae22d669fc7dd0f3f61130535a2670dedd))

### Breaking
* new changes ([`bb076da`](https://github.com/UBC-MDS/nameforme/commit/bb076dae22d669fc7dd0f3f61130535a2670dedd))

## v0.3.1 (2023-01-29)


## v0.3.0 (2023-01-29)
### Feature
* **new:** Updated readme. BREAKING CHANGE ([`4f60495`](https://github.com/UBC-MDS/nameforme/commit/4f6049568f08564a2bcf83919f3230f15da3cf9a))

## v0.2.0 (2023-01-29)
### Feature
* **new:** Updated project. BREAKING CHANGE 2 ([`e17ac61`](https://github.com/UBC-MDS/nameforme/commit/e17ac61d313bf0b6d1235915d1e80872d1d5b72e))

## v0.1.0 (2023-01-29)
### Feature
* **new:** Updated project. BREAKING CHANGE ([`9381a25`](https://github.com/UBC-MDS/nameforme/commit/9381a250eceda8f815d08ad18311201b4ba15a33))

## v0.0.20 (2023-01-29)


## v0.0.19 (2023-01-29)


## v0.0.18 (2023-01-29)


## v0.0.17 (2023-01-29)


## v0.0.16 (2023-01-29)


## v0.0.15 (2023-01-29)


## v0.0.14 (2023-01-29)


## v0.0.13 (2023-01-29)


## v0.0.12 (2023-01-28)


## v0.0.11 (2023-01-28)


## v0.0.10 (2023-01-28)


## v0.0.9 (2023-01-28)


## v0.0.8 (2023-01-28)


## v0.0.7 (2023-01-28)


## v0.0.6 (2023-01-28)


## v0.0.5 (2023-01-25)


## v0.0.4 (2023-01-25)


## v0.0.3 (2023-01-24)


## v0.1.0 (12/01/2023)

- First release of `nameforme`!

tests/test_find_name.py
from nameforme.nameforme import find_name

def test_find_name():
    """Test cases for find_name()"""
    
    """Test that the names are in the same length"""
    assert len(set(len(i) for i in find_name("F", "A", 3))) == 1, 'The list provides different length of names'
    
    """Test that the names match the required length"""
    assert next(iter(set(len(i) for i in find_name("F", "A", 3)))) == 3, 'The list does not provide the correct length of names'
    
    """Test that the names are start with the same letter"""
    assert len(set(i[0] for i in find_name("F", "A", 3))) == 1, 'The list provides different initial letter of names'
    
    """Test that the names match the required initial letter"""
    assert next(iter(set(i[0] for i in find_name("F", "A", 3)))) == 'A', 'The list does not provide the correct initial letter of names '
    
    """Test that there are no duplicate names in output"""
    assert len(set(find_name("F", "A", 3))) == len(find_name("F", "A", 3)), 'The list should not have duplicated names'
    

tests/test_nameforme.py
from nameforme.nameforme import find_name
from nameforme.nameforme import find_old_name
from nameforme.nameforme import find_similar_name
from nameforme.nameforme import find_unisex_name

import pandas as pd
import jellyfish as jf
import random
import pytest


def test_find_name():
    """Test cases for find_name()"""
    
    """Test the input types"""
    with pytest.raises(TypeError, match = "sex needs to be a str type."):
        find_name(1, "A", 3)
    
    with pytest.raises(TypeError, match = "init needs to be a str type."):
        find_name("F", 5, 3)
    
    with pytest.raises(TypeError, match = "length needs to be an int type."):
        find_name("F", "A", "a")
    
    """Test the input values"""
    with pytest.raises(Exception, match = "sex should be either 'F'/'f' or 'M'/'m'."):
        find_name("a", "A", 3) 
        
    with pytest.raises(Exception, match = "init should be a single character."):
        find_name("F", "qw", 3)   
        
    with pytest.raises(Exception, match = "init should be an alphabet."):
        find_name("F", "?", 3)
        
    with pytest.raises(Exception, match = "length should be larger than 0."):
        find_name("F", "A", 0)    
    
    """Test that the names are in the same length"""
    assert len(set(len(i) for i in find_name("F", "A", 3))) == 1, 'The list provides different length of names'
    
    """Test that the names match the required length"""
    assert next(iter(set(len(i) for i in find_name("F", "A", 3)))) == 3, 'The list does not provide the correct length of names'
    
    """Test that the names are start with the same letter"""
    assert len(set(i[0] for i in find_name("F", "A", 3))) == 1, 'The list provides different initial letter of names'
    
    """Test that the names match the required initial letter"""
    assert next(iter(set(i[0] for i in find_name("F", "A", 3)))) == 'A', 'The list does not provide the correct initial letter of names '
    
    """Test that there are no duplicate names in output"""
    assert len(set(find_name("F", "A", 3))) == 10, 'The list should not have duplicated names'
    
    """Test that the output list size"""
    assert len(find_name("F", "A", 1)) == 0, 'The list cannot has 10 output if the names does not exist'
    
    

def test_find_old_name():
    """Test cases for find_old_name()"""

    """Test that n names are output"""
    assert len(find_old_name('1980s', limit=3)) == 3, 'The list provides different length of names'
    
    """Test that there are no duplicate names in output"""
    assert len(set(find_old_name('1980s'))) == 10, 'The defualt list is not of length 10'
    
    """Test that repeated calls generate different names"""
    first_call = find_old_name('1980s', limit=10)
    second_call = find_old_name('1980s', limit=10)
    assert set(first_call) != set(second_call), 'The list generated is not random'

    """Test that repeated calls with seed given generate same names"""
    first_call = find_old_name('1980s', seed=123)
    second_call = find_old_name('1980s', seed=123)
    assert set(first_call) == set(second_call), 'The lists generated with given seed are not the same.'
    
    """Test that name is actually of given sex"""
    URL = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2022/2022-03-22/babynames.csv"
    data = pd.read_csv(URL)
    data = data[data["sex"] == "M"]["name"]
    out = find_old_name('1980s', sex='M')
    assert sum([ _ in list(data) for _ in find_old_name('1980s', sex='M')]) == len(out), "Does not give name of right sex."
    
def test_find_similar_name():
    """Test cases for find_similar_name()"""

    """Test that n names are output"""
    assert len(find_similar_name('Daniel', limit=3)) == 3, "Wrong number of names output"
    
    """Test that there are no duplicate names in output"""
    assert len(set(find_similar_name('Daniel'))) == 10, "Duplicate names in the output"
    
    """Test that repeated calls generate different names"""
    first_call = find_similar_name('Daniel', limit=20)
    second_call = find_similar_name('Daniel', limit=20)
    assert set(first_call) != set(second_call), "Sucessive calls don't seem random"

    """Crude similarity test: checking if first letter matches"""
    assert "D" in [name[0] for name in find_similar_name('Daniel', 3)], "Doesn't seem like the generated names are similar"
        
def test_find_unisex_name():
    """Test cases for find_unisex_name()"""
    
    """Test that the characters in the output are all alphabetic"""
    assert sum(str.isalpha(i) for i in find_unisex_name(bar=0.02,limit=10)) == 10, 'The output is illegal'
    
    """Test that the length of the output is correct"""
    assert len(find_unisex_name(bar=0.02,limit=5)) == 5, 'The list does not provide the correct length of names'
    
    """Test that no duplicate names are found in the output"""
    assert len(set(find_unisex_name(bar=0.02,limit=10))) == 10, 'The length of the output is not correct'
    
    """Test that changing bar value create new name list"""
    assert set(find_unisex_name(bar=0.02,limit=10)) != set(find_unisex_name(bar=0.01,limit=10)), 'The output of different bar value should not be the same '
    
    
    

tests/test_find_old_name.py
# Authous: Daniel Cairns, Eyre Hong, Bruce Wu, Zilong Yi (UBC MDS)
# Date: Jan 20th, 2023

from nameforme.nameforme import find_old_name
import pandas as pd
from numpy import sum

def test_find_old_name():
    """Test cases for find_old_name()"""

    """Test that n names are output"""
    assert len(find_old_name('1980s', limit=3)) == 3, 'The list provides different length of names'
    
    """Test that there are no duplicate names in output"""
    assert len(set(find_old_name('1980s'))) == 10, 'The defualt list is not of length 10'
    
    """Test that repeated calls generate different names"""
    first_call = find_old_name('1980s', limit=10)
    second_call = find_old_name('1980s', limit=10)
    assert set(first_call) != set(second_call), 'The list generated is not random'

    """Test that repeated calls with seed given generate same names"""
    first_call = find_old_name('1980s', seed=123)
    second_call = find_old_name('1980s', seed=123)
    assert set(first_call) != set(second_call), 'The lists generated with given seed are not the same.'
    
    """Test that name is actually of given sex"""
    URL = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2022/2022-03-22/babynames.csv"
    data = pd.read_csv(URL)
    data = data[data["sex"] == "M"]
    out = find_old_name('1980s', sex='M')
    assert sum([ _ in list(data) for _ in find_old_name('1980s', sex='M')]) == len(out), "Does not give name of right sex."

src/nameforme/find_name.py
# Authours: Daniel Cairns, Eyre Hong, Bruce Wu, Zilong Yi (UBC MDS)
# Date: Jan 14th, 2023

import pandas as pd
import random

def find_name(sex, init, length):
    """
    Generate the a random set of 10 suggested baby names based on the given limitations.
    Parameters
    ----------
    sex: string, 
        The sex of baby's name, either 'F' or 'M' or 'f' or 'm'
    init : string
        The initial of baby's name, a single alphabet character
    length : int
        The length of baby's name, larager than 0
    Returns
    -------
    list:
        A name list containing random suggested names based on the given limitation.
    Examples
    --------
    >>> find_name('F', 'A', 3)
    >>> ['Ada', 'Aja', 'Ani', 'Aya', 'Ava', 'Ana', 'Ari', 'Ali', 'Ami', 'Amy']
    """
    # Check input type of sex
    if not isinstance(sex, str):
        raise TypeError("sex needs to be a str type.")
    
    # Check input type of init
    if not isinstance(init, str):
        raise TypeError("init needs to be a str type.")

    # Check input type of length
    if not isinstance(length, int):
        raise TypeError("length needs to be an int type.")
    
    # Check input value of sex
    if not (sex == 'F' or sex == 'M' or sex == 'f' or sex == 'm'):
        raise Exception("sex should be either 'F/f' or 'M/m'")
    
    # Check input value of init
    if not (len(init) == 1):
        raise Exception("init should be a single character")
    
    if not (str.isalpha(init)):
        raise Exception("init should be an alphabet")
    
    # Check input value of length
    if length <= 0:
        raise Exception("length should be larger than 0")
    
    # Data loading and cleaning   
    url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2022/2022-03-22/babynames.csv"
    raw_df = pd.read_csv(url)
    raw_df = raw_df[raw_df['n'] >= 100] # Keep only names that had at least 100 births for a single gender in a single year
    
    # Filter data based on the arguments
    df_sex = raw_df.loc[raw_df['sex'] == sex.capitalize()]
    df_init = df_sex.loc[df_sex['name'].str.startswith(init.capitalize(), na=False)]
    df_len = df_init.loc[df_init['name'].str.len() == length]
    
    # Create name list and randomly select 10 names
    name_list = df_len['name'].unique().tolist()
    if len(name_list) < 10:
        sampled = random.sample(name_list, k=len(name_list))
    else:
        sampled = random.sample(name_list, k=10)    
    
    return sampled

src/nameforme/find_old_name.py
# Authours: Daniel Cairns, Eyre Hong, Bruce Wu, Zilong Yi (UBC MDS)
# Date: Jan 14th, 2023

def find_old_name(tp,limit=10, sex="uni", seed=None):
    """
    Generate a random set of 10(if there are that many exists) suggested neutral(by default) baby names 
    based on the given time period and sex.
    Parameters
    ----------
    tp : string, default 10
        A string specifying time period from 1880 to 2018
        List of possible qualified inputs:
        ['1880s','1890s','1900s','1910s','1920s','1930s','1940s','1950s','1960s','1970s','1980s','1990s','2000s','2010s']
    limit : float
        The number of name in the output list. 
    sex : string, default "uni"
        The aimed sex of the name in the ouput.
        List of possible qualified inputs:
        ["uni","M","F] 
    seed : int
        if provided, returns the same random set. 
    Returns
    -------
    list:
        A name list containing random suggested old names based on the given limitation.
    Examples
    --------
    >>> find_old_name('1980s', seed=123)
    >>> ['Neng','Rashida','Krister','Garry','Kai','Chezare','Dolores','Tyler','Skylar','Tyrie']
    """
    # Load data
    import pandas as pd
    import numpy as np
    URL = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2022/2022-03-22/babynames.csv"
    data = pd.read_csv(URL)
    # Data wrangling
    data["tp"] = np.select(
        [
            data["year"].between(1880,1890, inclusive="left"),
            data["year"].between(1890,1900, inclusive="left"),
            data["year"].between(1900,1910, inclusive="left"),
            data["year"].between(1910,1920, inclusive="left"),
            data["year"].between(1920,1930, inclusive="left"),
            data["year"].between(1930,1940, inclusive="left"),
            data["year"].between(1940,1950, inclusive="left"),
            data["year"].between(1950,1960, inclusive="left"),
            data["year"].between(1960,1970, inclusive="left"),
            data["year"].between(1970,1980, inclusive="left"),
            data["year"].between(1980,1990, inclusive="left"),
            data["year"].between(1990,2000, inclusive="left"),
            data["year"].between(2000,2010, inclusive="left"),
            data["year"].between(2010,2020, inclusive="left"),
        ],
        [
            '1880s',
            '1890s',
            '1900s',
            '1910s',
            '1920s',
            '1930s',
            '1940s',
            '1950s',
            '1960s',
            '1970s',
            '1980s',
            '1990s',
            '2000s',
            '2010s'
        ]
    )
    # Setting seed.
    if seed is not None:
        np.random.seed(seed)
    if tp not in ['1880s','1890s','1900s','1910s','1920s','1930s','1940s','1950s','1960s','1970s','1980s','1990s','2000s','2010s'] or sex not in ["uni",'M','F']:
        raise Exception("Sorry, please enter valid time periods/sex!")
    
    df = data[data["tp"]==tp]
    if sex == "uni":
        F = df[df["sex"] == 'F']["name"]
        M = df[df["sex"] == "M"]["name"]
        uni_df = list(set(F).intersection(set(M)))
        if len(uni_df) < 10:
            return uni_df
        else:
            return np.random.choice(uni_df, 10, replace=False).tolist()
    else:
        df = df[df["sex"] == sex]["name"]
        return np.random.choice(df,10,replace=False).tolist()

src/nameforme/find_unisex_name.py
# Authours: Daniel Cairns, Eyre Hong, Bruce Wu, Zilong Yi (UBC MDS)
# Date: Jan 14th, 2023

import pandas as pd
import numpy as np

def find_unisex_name(bar,limit=10):
    """
    Generate the a random set of 10 suggested neutral baby names 
    based on the given limitation and baby names in the past years.
    Parameters
    ----------
    limit : float
        A float controls the minimum proportion of the neutral names in a single year in the dataframe.
    count : integer
        The length of the output with a default value of 10
    Returns
    -------
    list:
        A word list containing random suggested neutral names based on the given limitation
        If the length of the word list is less than the limitation, it will return all the names in the word list
    Examples
    --------
    >>> limit=0.0001
    >>> find_unisex_name(limit)
    >>> ['Skylar', 'Azariah', 'Royal', 'Hayden', 'Emerson', 'Rowan', 'Baylor', 'Dakota', 'River', 'Emory']
    """
    
    #check input value
    if not (bar > 0 and bar < 1):
        raise Exception("bar value should within 0 and 1")
    
    # Load data
    url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2022/2022-03-22/babynames.csv"
    data = pd.read_csv(url)
    
    # data wrangling
    bar_data_boy = data[(data['prop'] > bar) & (df['sex'] == 'M')] # select all the boy names with prop greater than bar
    bar_data_girl = data[(data['prop'] > bar) & (df['sex'] == 'F')] # select all the girl names with prop greater than bar
    target_names = []
    for girl_name in set(bar_data_girl['name']):
        if girl_name in set(bar_data_boy['name']):
            target_names.append(girl_name)
    
    # Create name list
    if len(target_names) = 0:
        raise Exception("Please lower the bar value")
    if len(target_names) < limit:
        result = target_names
    else:
        result = np.random.choice(target_names, limit, replace=False).tolist()
    
    return result


src/nameforme/find_similar_name.py
# Authours: Daniel Cairns, Eyre Hong, Bruce Wu, Zilong Yi (UBC MDS)
# Date: Jan 14th, 2023

import pandas as pd
import jellyfish as jf

def find_similar_name(match_name, limit=10):
    """
    Generate a random list of names that sound similar to a given user input name.
    
    Uses Match rating approach algorithm to determine similarity
      
    Parameters
    ----------
    match_name : string
        Name for comparison; output names will sound like this one 
    limit : int
        The number of names in the output list. (default = 10)
    Returns
    -------
    list:
        A name list containing random suggested similar names based on the given name.
    Examples
    --------
    >>> find_similar_name('Elizabeth', 5)
    >>> ['Elisabeth', 'Elliott', 'Ellsworth', 'Emmalynn', 'Ryley']
    """
    
    # Check inputs
    if not isinstance(match_name, str):
        raise TypeError("input name must be a string")

    if not isinstance(limit, int):
        raise TypeError("Limit must be an integer")
    
    # Data loading and cleaning
    url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2022/2022-03-22/babynames.csv"
    raw_df = pd.read_csv(url)
    raw_df = raw_df[raw_df["n"] >= 100] # Keep only names that had at least 100 births for a single gender in a single year
    unique_names = raw_df['name'].unique()
    unique_names = unique_names[unique_names != match_name] # Remove target name from potential outputs
    
    # Calculate targets 
    encodings = [jf.match_rating_codex(n) for n in unique_names]
    target = jf.match_rating_codex(match_name)
    distance = [jf.jaro_winkler_similarity(target, e) for e in encodings]
    
    df = pd.DataFrame(data = [unique_names, encodings, distance]).T
    df.columns = ["name", "encoding", "match_score"]
    df["weight"] = (df["match_score"]*2)**10 # Heavily weigh higher scores
    
    # Sample n names to output, weighted by match_score
    sampled = df.sample(n = limit, weights = "weight")
    sampled = sampled.sort_values(by = "match_score", ascending = False)
    return(sampled["name"].to_list())


src/nameforme/nameforme.py
# Authours: Daniel Cairns, Eyre Hong, Bruce Wu, Zilong Yi (UBC MDS)
# Date: Jan 14th, 2023

import pandas as pd
import numpy as np
import jellyfish as jf
import random

def find_name(sex, init, length):
    """
    Generate the a random set of 10 suggested baby names based on the given limitations.
    Parameters
    ----------
    sex: string, 
        The sex of baby's name, either 'F' or 'M' or 'f' or 'm'
    init : string
        The initial of baby's name, a single alphabet character
    length : int
        The length of baby's name, larager than 0
    Returns
    -------
    list:
        A name list containing random suggested names based on the given limitation.
    Examples
    --------
    >>> find_name('F', 'A', 3)
    >>> ['Ada', 'Aja', 'Ani', 'Aya', 'Ava', 'Ana', 'Ari', 'Ali', 'Ami', 'Amy']
    """
    # Check input type of sex
    if not isinstance(sex, str):
        raise TypeError("sex needs to be a str type.")
    
    # Check input type of init
    if not isinstance(init, str):
        raise TypeError("init needs to be a str type.")

    # Check input type of length
    if not isinstance(length, int):
        raise TypeError("length needs to be an int type.")
    
    # Check input value of sex
    if not (sex == 'F' or sex == 'M' or sex == 'f' or sex == 'm'):
        raise Exception("sex should be either 'F'/'f' or 'M'/'m'.")
    
    # Check input value of init
    if not (len(init) == 1):
        raise Exception("init should be a single character.")
    
    if not (str.isalpha(init)):
        raise Exception("init should be an alphabet.")
    
    # Check input value of length
    if length <= 0:
        raise Exception("length should be larger than 0.")
    
    # Data loading and cleaning   
    url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2022/2022-03-22/babynames.csv"
    raw_df = pd.read_csv(url)
    raw_df = raw_df[raw_df['n'] >= 100] # Keep only names that had at least 100 births for a single gender in a single year
    
    # Filter data based on the arguments
    df_sex = raw_df.loc[raw_df['sex'] == sex.capitalize()]
    df_init = df_sex.loc[df_sex['name'].str.startswith(init.capitalize(), na=False)]
    df_len = df_init.loc[df_init['name'].str.len() == length]
    
    # Create name list and randomly select 10 names
    name_list = df_len['name'].unique().tolist()
    if len(name_list) < 10:
        sampled = random.sample(name_list, k=len(name_list))
    else:
        sampled = random.sample(name_list, k=10)    
    
    return sampled


def find_old_name(tp,limit=10, sex="uni", seed=None):
    """
    Generate a random set of 10(if there are that many exists) suggested neutral(by default) baby names 
    based on the given time period and sex.
    Parameters
    ----------
    tp : string, default 10
        A string specifying time period from 1880 to 2018
        List of possible qualified inputs:
        ['1880s','1890s','1900s','1910s','1920s','1930s','1940s','1950s','1960s','1970s','1980s','1990s','2000s','2010s']
    limit : float
        The number of name in the output list. 
    sex : string, default "uni"
        The aimed sex of the name in the ouput.
        List of possible qualified inputs:
        ["uni","M","F] 
    seed : int
        if provided, returns the same random set. 
    Returns
    -------
    list:
        A name list containing random suggested old names based on the given limitation.
    Examples
    --------
    >>> find_old_name('1980s', seed=123)
    >>> ['Neng','Rashida','Krister','Garry','Kai','Chezare','Dolores','Tyler','Skylar','Tyrie']
    """
    # Load data
    URL = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2022/2022-03-22/babynames.csv"
    data = pd.read_csv(URL)
    # Data wrangling
    data["tp"] = np.select(
        [
            data["year"].between(1880,1890, inclusive="left"),
            data["year"].between(1890,1900, inclusive="left"),
            data["year"].between(1900,1910, inclusive="left"),
            data["year"].between(1910,1920, inclusive="left"),
            data["year"].between(1920,1930, inclusive="left"),
            data["year"].between(1930,1940, inclusive="left"),
            data["year"].between(1940,1950, inclusive="left"),
            data["year"].between(1950,1960, inclusive="left"),
            data["year"].between(1960,1970, inclusive="left"),
            data["year"].between(1970,1980, inclusive="left"),
            data["year"].between(1980,1990, inclusive="left"),
            data["year"].between(1990,2000, inclusive="left"),
            data["year"].between(2000,2010, inclusive="left"),
            data["year"].between(2010,2020, inclusive="left"),
        ],
        [
            '1880s',
            '1890s',
            '1900s',
            '1910s',
            '1920s',
            '1930s',
            '1940s',
            '1950s',
            '1960s',
            '1970s',
            '1980s',
            '1990s',
            '2000s',
            '2010s'
        ]
    )
    # Setting seed.
    if seed is not None:
        np.random.seed(seed)
    if tp not in ['1880s','1890s','1900s','1910s','1920s','1930s','1940s','1950s','1960s','1970s','1980s','1990s','2000s','2010s'] or sex not in ["uni",'M','F']:
        raise Exception("Sorry, please enter valid time periods/sex!")
    
    df = data[data["tp"]==tp]
    if sex == "uni":
        F = df[df["sex"] == 'F']["name"]
        M = df[df["sex"] == "M"]["name"]
        uni_df = list(set(F).intersection(set(M)))
        if len(uni_df) < 10:
            return uni_df
        else:
            return np.random.choice(uni_df, limit, replace=False).tolist()
    else: 
        df = df[df["sex"] == sex]["name"]
        return np.random.choice(df,limit,replace=False).tolist()
    

def find_similar_name(match_name, limit=10):
    """
    Generate a random list of names that sound similar to a given user input name.
    
    Uses Match rating approach algorithm to determine similarity
      
    Parameters
    ----------
    match_name : string
        Name for comparison; output names will sound like this one 
    limit : int
        The number of names in the output list. (default = 10)
    Returns
    -------
    list:
        A name list containing random suggested similar names based on the given name.
    Examples
    --------
    >>> find_similar_name('Elizabeth', 5)
    >>> ['Elisabeth', 'Elliott', 'Ellsworth', 'Emmalynn', 'Ryley']
    """
    
    # Check inputs
    if not isinstance(match_name, str):
        raise TypeError("input name must be a string")

    if not isinstance(limit, int):
        raise TypeError("Limit must be an integer")
    
    # Data loading and cleaning
    url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2022/2022-03-22/babynames.csv"
    raw_df = pd.read_csv(url)
    raw_df = raw_df[raw_df["n"] >= 100] # Keep only names that had at least 100 births for a single gender in a single year
    unique_names = raw_df['name'].unique()
    unique_names = unique_names[unique_names != match_name] # Remove target name from potential outputs
    
    # Calculate targets 
    encodings = [jf.match_rating_codex(n) for n in unique_names]
    target = jf.match_rating_codex(match_name)
    distance = [jf.jaro_winkler_similarity(target, e) for e in encodings]
    
    df = pd.DataFrame(data = [unique_names, encodings, distance]).T
    df.columns = ["name", "encoding", "match_score"]
    df["weight"] = (df["match_score"]*2)**10 # Heavily weigh higher scores
    
    # Sample n names to output, weighted by match_score
    sampled = df.sample(n = limit, weights = "weight")
    sampled = sampled.sort_values(by = "match_score", ascending = False)
    return(sampled["name"].to_list())

def find_unisex_name(bar,limit=10):
    """
    Generate the a random set of 10 suggested neutral baby names 
    based on the given limitation and baby names in the past years.
    Parameters
    ----------
    limit : float
        A float controls the minimum proportion of the neutral names in a single year in the dataframe.
    count : integer
        The length of the output with a default value of 10
    Returns
    -------
    list:
        A word list containing random suggested neutral names based on the given limitation
        If the length of the word list is less than the limitation, it will return all the names in the word list
    Examples
    --------
    >>> limit=10
    >>> find_unisex_name(0.2,limit)
    >>> ['Skylar', 'Azariah', 'Royal', 'Hayden', 'Emerson', 'Rowan', 'Baylor', 'Dakota', 'River', 'Emory']
    """
    
    #check input value
    if not (bar > 0 and bar < 1):
        raise Exception("bar value should within 0 and 1")
    
    # Load data
    url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2022/2022-03-22/babynames.csv"
    data = pd.read_csv(url)
    
    # data wrangling
    bar_data_boy = data[(data['prop'] > bar) & (data['sex'] == 'M')] # select all the boy names with prop greater than bar
    bar_data_girl = data[(data['prop'] > bar) & (data['sex'] == 'F')] # select all the girl names with prop greater than bar
    target_names = []
    for girl_name in set(bar_data_girl['name']):
        if girl_name in set(bar_data_boy['name']):
            target_names.append(girl_name)
    
    # Create name list
    common_unisex_names=['Skylar', 'Azariah', 'Royal', 'Hayden', 'Emerson', 
                         'Rowan', 'Baylor', 'Dakota', 'River', 'Emory','Jessie',
                         'Marion','Jackie','Alva','Ollie','Jodie','Cleo','Kerry']
    if len(target_names) < limit:
        result = target_names + np.random.choice(common_unisex_names, limit-len(target_names), replace=False).tolist()
    else:
        result = np.random.choice(target_names, limit, replace=False).tolist()
    
    return result